astropy/table/table.py

```diff
@@ -775,8 +778,13 @@ def _init_from_list_of_dicts(self, data, names, dtype, n_cols, copy):
                     cols[name].append(row[name])
                 except KeyError:
                     raise ValueError('Row {0} has no value for column {1}'.format(i, name))
+
         if all(name is None for name in names):
-            names = sorted(names_from_data)
```

Comment:
With the other change suggested then this just becomes as below (since `names_from_data` is already in the desired order):
```
if all(name is None for name in names):
    names = names_from_data
```